### PR TITLE
refactor: Introduce EvictionManager and use it for the inbound eviction logic

### PR DESCRIPTION
**Pull request by dergoegge**
This PR splits off the next couple commits from #25268 that introduce the `EvictionManager` and use it for  the inbound eviction logic.

One instance of the `EvictionManager` is created at start up and passed as a reference to the connection and peer managers. The connection and peer managers report all eviction relevant information (for inbound connections) to the eviction manager who ultimately suggests nodes to evict as the result of `EvictionManager::SelectNodeToEvict`.

### CODE REVIEW

File: src/node/eviction.cpp [hunk 1]
```diff
@@ -243,6 +244,53 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
...
 
+void EvictionManagerImpl::AddCandidate(NodeId id, std::chrono::seconds connected,
+                                       uint64_t keyed_net_group, bool prefer_evict,
+                                       bool is_local, Network network,
+                                       bool noban, ConnectionType conn_type)
+{
+    LOCK(m_candidates_mutex);
+    // The default values used to initialise the eviction candidates result in
+    // newer candidates being more likely to be evicted.
+    NodeEvictionCandidate candidate{
```
Comment (theuni) [inline src/node/eviction.cpp:256]: I wonder if it might make sense to split NodeEvictionCandidate into 2 parts: the const properties that don't change (like conn_type/noban) and the non-const state that's intended to be/might be updated?

Maybe something like:

```c++
class NodeEvictionCandidate {
  class Properties {
    NodeId m_id;
    ConnectionType m_conn_type;
    bool m_noban;
    ...
  };
  NodeEvictionCandidate(NodeProperties properties) : m_properties(properties){}
private:
  const NodeProperties m_properties;
  bool m_relay_txs;
  bool fBloomFilter;
  ...
};
...
void EvictionManagerImpl::AddCandidate(NodeEvictionCandidate::Properties props) {...}
```

I think that might be more straightforward to use as it's self-documenting, while also allowing for safe lock-free access to the stuff that can't change.



Maybe that's easier said than done though. Thoughts?

**Review by theuni — COMMENTED** [review]
Concept ACK.

Added a high-level question about `NodeEvictionCandidate` to kick things off, will continue with review.

File: src/net.cpp [hunk 2]
```diff
@@ -1112,6 +1095,11 @@ void CConnman::DisconnectNodes()
             }
         }
     }
+
+    for (NodeId id : disconnected_eviction_candidates) {
+        m_evictionman.RemoveCandidate(id);
+    }
```
Comment (jnewbery) [inline src/net.cpp:1101]: Is there a reason not to do this in the `Delete disconnected nodes` loop below?

File: src/net.h [hunk 3]
```diff
@@ -582,7 +554,6 @@ class CNode
     /** A ping-pong round trip has completed successfully. Update latest and minimum ping times. */
```
Comment (jnewbery) [inline src/net.h:554]:
```suggestion
    /** A ping-pong round trip has completed successfully. Update latest ping time. */
```

In fact, `m_last_ping_time` may as well be moved into `Peer` since it's not used in any net level logic and is only for RPC/GUI stats. Could be done as a follow-up.

File: src/net.cpp [hunk 4]
```diff
@@ -2555,6 +2554,22 @@ void CConnman::GetNodeStats(std::vector<CNodeStats>& vstats) const
     for (CNode* pnode : m_nodes) {
         vstats.emplace_back();
         pnode->CopyStats(vstats.back());
+
+        auto min_ping_time{m_evictionman.GetMinPingTime(pnode->GetId())};
+        if (min_ping_time) {
```
Comment (jnewbery) [inline src/net.cpp:2559]: Can be shortened to

```suggestion
        if (auto min_ping_time = m_evictionman.GetMinPingTime(pnode->GetId())) {
```

if you think that's better. Same below.

File: src/node/eviction.h [hunk 5]
```diff
@@ -15,55 +15,53 @@
...
+     * distinct, difficult-to-forge characteristics. The protection logic picks out
+     * fixed numbers of desirable peers per various criteria, followed by (mostly)
+     * ratios of desirable or disadvantaged peers. If any eviction candidates
+     * remain, the selection logic chooses a peer to evict.
+     */
+    [[nodiscard]] std::optional<NodeId> SelectNodeToEvict() const;
+
+    /** A ping-pong round trip has completed successfully. Update minimum ping time. */
+    void UpdateMinPingTime(NodeId id, std::chrono::microseconds ping_time);
+    std::optional<std::chrono::microseconds> GetMinPingTime(NodeId id) const;
```
Comment (jnewbery) [inline src/node/eviction.h:46]: Once the outbound eviction logic has also been moved into EvictionManager, can these three getters be combined into a single `GetEvictionCandidateStats()` or similar, since they'll only be used for reporting stats to the user. If you think that's a good idea, perhaps add that to the function comment for now.

File: src/node/eviction.cpp [hunk 6]
```diff
@@ -238,3 +241,194 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
...
+    {
+        LOCK(m_candidates_mutex);
+        for (const auto& [id, candidate] : m_candidates) {
+            candidates.push_back(candidate);
+        }
+    }
+    return ::SelectNodeToEvict(std::move(candidates));
+}
+
+void EvictionManagerImpl::UpdateMinPingTime(NodeId id, std::chrono::microseconds ping_time)
```
Comment (jnewbery) [inline src/node/eviction.cpp:293]: /me idly wonders whether these setters can be templated to avoid repetitive boilerplate code

File: src/test/net_peer_eviction_tests.cpp [hunk 7]
```diff
@@ -2,6 +2,7 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <node/eviction_impl.h>
```
Comment (jnewbery) [inline src/test/net_peer_eviction_tests.cpp:5]: sort

File: src/node/eviction_impl.h [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    bool fBloomFilter;
+    uint64_t nKeyedNetGroup;
+    bool prefer_evict;
+    bool m_is_local;
+    Network m_network;
+    bool m_noban;
+    ConnectionType m_conn_type;
+};
+
+[[nodiscard]] std::optional<NodeId> SelectNodeToEvict(std::vector<NodeEvictionCandidate>&& vEvictionCandidates);
```
Comment (jnewbery) [inline src/node/eviction_impl.h:36]: Is there any reason why this and `ProtectEvictionCandidatesByRatio()` aren't members of `EvictionManagerImpl`? I'm not saying they have to be, but I'm interested in the design choice.

File: src/test/fuzz/node_eviction.cpp [hunk 9]
```diff
@@ -2,6 +2,7 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <node/eviction_impl.h>
```
Comment (jnewbery) [inline src/test/fuzz/node_eviction.cpp:5]: sort

File: src/node/eviction.cpp (hunk 1 above)
Comment (jnewbery) [reply src/node/eviction.cpp:256]: Even if you don't do this, I'd suggest marking the const members const, and group them before the mutable members.

**Review by jnewbery — COMMENTED** [review]
Big concept ACK!

Minor cosmetic comments inline.

File: src/net.cpp (hunk 2 above)
Comment (theuni) [reply src/net.cpp:1101]: It's a different set of nodes, so different loop. Do you mean to do it in the same scope?

Comment (jnewbery) [reply src/net.cpp:1101]: I guess I don't understand why the node can't be removed from the EvictionManager either:

- in the loop above, at the point where the node is removed from `m_nodes` and added to `m_nodes_disconnected`; or
- in the loop below, at the point where the node is removed from `m_nodes_disconnected`.

Perhaps a code comment could be added to explain why.

File: src/net.h (hunk 3 above)
Comment (dergoegge) [reply src/net.h]: Gonna leave this for a follow-up.

File: src/net.cpp (hunk 4 above)
Comment (dergoegge) [reply src/net.cpp]: Done.

File: src/node/eviction.cpp (hunk 1 above)
Comment (dergoegge) [reply src/node/eviction.cpp:255]: I tried to do this but the `src/test/net_peer_eviction_tests.cpp` tests sadly rely on no members of `NodeEvictionCandidate` being const. Would like to leave this for a follow up, since it requires changing those tests quite a bit.

File: src/net.cpp (hunk 2 above)
Comment (dergoegge) [reply src/net.cpp:1100]: Cory originally suggested to it this way to untangle the removal from `m_nodes_mutex`. (See https://github.com/bitcoin/bitcoin/pull/25268#discussion_r903003875)

I think it could make sense to do it in the `Delete disconnected nodes` loop below which is also independent of `m_node_mutex`. Another option would be to do it in `PeerManagerImpl::FinalizeNode()` but that might be weird since we add candidates in `net`.

File: src/node/eviction_impl.h (hunk 8 above)
Comment (dergoegge) [reply src/node/eviction_impl.h:36]: This is something that @theuni  originally also had on his branch and i just went with the same design. To me this seems good because some of the tests purely test these functions without needing the eviction manager and adding them as members would require more test refactoring.

Comment (jnewbery) [reply src/node/eviction_impl.h:36]: I think that test refactoring would probably be a good thing, since it would ensure that the unit test is not testing a state that EvictionManager can't get into in the product code. In general, that seems like a good model - unit tests use the public methods to update the state of the object being tested, but optionally also have access to const test-only methods to peek inside the state of the object if it's difficult to expose that state in other ways.

Certainly not required in this PR. Could be done in a follow-up (or not at all).

File: src/net.cpp (hunk 2 above)
Comment (dergoegge) [reply src/net.cpp]: Moved this back into the loop above.

File: src/node/eviction.cpp (hunk 6 above)
Comment (dergoegge) [reply src/node/eviction.cpp:292]: Tried to come up with a cool way of doing this with templates but didn't end with anything useful.

**Comment by dergoegge** [comment]
Rebased.

File: src/node/eviction.cpp (hunk 1 above)
Comment (theuni) [reply src/node/eviction.cpp:255]: Hmm, I'd say if the tests rely on non-const values, then the tests need to be reworked somewhat (I assume they're just re-using existing candidates as opposed to creating new ones for the sake of simplicity). Either that or they're testing state-changes that aren't possible.

Agree that it's not worth refactoring the tests in this PR, but could you maybe organize the members and add a note to the definition of `NodeEvictionCandidate` to make the possible state changes clear? I *think* there are basically 3 types, though I may be oversimplifying:
- These never change after creation
- These may change once
- These may change at any time

**Comment by theuni** [comment]
Needs a quick linter fix.

**Comment by dergoegge** [comment]
This has been rebased (including a small linter fix) and is ready for review again!

File: src/node/eviction.cpp [hunk 10]
```diff
@@ -243,6 +244,53 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
 EvictionManagerImpl::EvictionManagerImpl() {}
 EvictionManagerImpl::~EvictionManagerImpl() = default;
 
+void EvictionManagerImpl::AddCandidate(NodeId id, std::chrono::seconds connected,
+                                       uint64_t keyed_net_group, bool prefer_evict,
+                                       bool is_local, Network network,
+                                       bool noban, ConnectionType conn_type)
+{
+    LOCK(m_candidates_mutex);
+    // The default values used to initialise the eviction candidates result in
```
Comment (naumenkogs) [inline src/node/eviction.cpp:253]: Why don't go further and summarize which parameters actually matter? Or just mention `why` it is so

**Comment by naumenkogs** [comment]
Concept ACK, light code review ACK 29c853614ac7ffbd6ab6c009f297f92149843e32.
This is a great refactor.

File: src/test/denialofservice_tests.cpp [hunk 11]
```diff
@@ -45,10 +45,19 @@ BOOST_FIXTURE_TEST_SUITE(denialofservice_tests, TestingSetup)
 // work.
 BOOST_AUTO_TEST_CASE(outbound_slow_chain_eviction)
 {
-    ConnmanTestMsg& connman = static_cast<ConnmanTestMsg&>(*m_node.connman);
-    // Disable inactivity checks for this test to avoid interference
-    connman.SetPeerConnectTimeout(99999s);
-    PeerManager& peerman = *m_node.peerman;
+    auto evictionman = std::make_unique<EvictionManager>();
```
Comment (theuni) [inline src/test/denialofservice_tests.cpp:48]: Why is this switched away from using `m_node`? Not that I object, it seems reasonable to me.

If it's just because "it's the only use of `m_node.connman` in these tests", maybe break that change out as a separate commit?

**Comment by dergoegge** [comment]
Rebased and added some docs for the initial `NodeEvictionCandidate` values.

File: src/node/eviction.cpp (hunk 10 above)
Comment (dergoegge) [reply src/node/eviction.cpp:253]: Added some comments for these values.

File: src/node/eviction.cpp (hunk 6 above)
Comment (ajtowns) [reply src/node/eviction.cpp:301]: I think something like this would work:

```c++
class EvictionManagerImpl
{
    ...
    template <typename Fn>
    auto CandidateOp(NodeId id, Fn&& fn)
    {
        LOCK(m_candidates_mutex);
        auto it = m_candidates.find(offset);
        if (it != m_candidates.end()) {
            return fn(it->second);
        } else {
            // return a dummy value of the right type, including void
            return ([]() -> decltype(fn(it->second)) { return {}; })();
        }
    }
    ...
};

void EvictionManager::UpdateLastBlockTime(NodeId id, std::chrono::seconds block_time)
{
    m_impl->CandidateOp(id, [](NodeEvictionCandidate& nec) {
        nec.m_last_block_time = block_time;
    });
}

void EvictionManager::GetLastBlockTime(NodeId id) const
{
    return m_impl->CandidateOp(id, [](const NodeEvictionCandidate& nec) {
        return nec.m_last_block_time;
    });
}
```

File: src/net.cpp [hunk 12]
```diff
@@ -2606,6 +2601,19 @@ void CConnman::GetNodeStats(std::vector<CNodeStats>& vstats) const
...
+        if (auto min_ping_time = m_evictionman.GetMinPingTime(pnode->GetId())) {
+            vstats.back().m_min_ping_time = *min_ping_time;
+        }
+
+        if (auto last_block_time = m_evictionman.GetLastBlockTime(pnode->GetId())) {
+            vstats.back().m_last_block_time = *last_block_time;
+        }
+
+        if (auto last_tx_time = m_evictionman.GetLastTxTime(pnode->GetId())) {
+            vstats.back().m_last_tx_time = *last_tx_time;
```
Comment (ajtowns) [inline src/net.cpp:2614]: `m_evictionman.CopyEvictionStats(pnode->GetId(), vstats.back());` ?

File: src/net_processing.cpp [hunk 13]
```diff
@@ -5008,7 +5015,7 @@ void PeerManagerImpl::EvictExtraOutboundPeers(std::chrono::seconds now)
             if (pnode->GetId() > youngest_peer.first) {
                 next_youngest_peer = youngest_peer;
                 youngest_peer.first = pnode->GetId();
-                youngest_peer.second = pnode->m_last_block_time;
+                youngest_peer.second = *Assert(m_evictionman.GetLastBlockTime(pnode->GetId()));
```
Comment (ajtowns) [inline src/net_processing.cpp:5018]: Repeatedly looking up the eviction manager map and locking and unlocking access to it, and introducing the potential for maps to be out of sync and having to add asserts to check for that doesn't seem ideal...

File: src/net_processing.cpp [hunk 14]
```diff
@@ -490,7 +491,7 @@ struct CNodeState {
 class PeerManagerImpl final : public PeerManager
 {
 public:
-    PeerManagerImpl(CConnman& connman, AddrMan& addrman,
+    PeerManagerImpl(CConnman& connman, AddrMan& addrman, EvictionManager& evictionman,
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/net_processing.cpp:494]: I think it makes sense to not have the eviction manager be owned by another component, in the context of also moving the outbound eviction logic to the eviction manager. I see it as more of a standalone component like the addrman or banman.

This might also give us flexibility in mocking it if we wanted to for testing purposes (e.g. to test that net processing and net are updating the stats correctly, or have a `FuzzedEvictionManager` to let the fuzzer choose which peers to evict).

File: src/net.cpp (hunk 12 above)
Comment (dergoegge) [reply src/net.cpp:2614]: Yea something like that would make sense. Also similar to John's [comment](https://github.com/bitcoin/bitcoin/pull/25572#discussion_r919919355).

**Review by dergoegge — COMMENTED** [review]
@ajtowns Thank you for the comments!

File: src/net.cpp [hunk 15]
```diff
@@ -895,20 +895,7 @@ size_t CConnman::SocketSendData(CNode& node) const
  */
 bool CConnman::AttemptToEvictConnection()
 {
-    std::vector<NodeEvictionCandidate> vEvictionCandidates;
-    {
-        LOCK(m_nodes_mutex);
-        for (const CNode* node : m_nodes) {
-            if (node->fDisconnect)
```
Comment (mzumsande) [inline src/net.cpp:907]: This removes the check `node->fDisconnect`, which I think could be dangerous:
We set `fDisconnect` in `AttemptToEvictConnection()`, but don't immediately disconnect. The actual cleanup happens at some later point, when `ThreadSocketHandler` calls `DisconnectNodes()`.
So if we have an attacker that makes new connections to us rapidly, there would be a race between `ThreadSocketHandler` and the attacker making the next connection to us. If the attacker wins this race, we could accept two inbound connections, but, without the check that is being removed here, `AttemptToEvictConnection()` could mark the same inbound peer for eviction twice, so that we'll actually only evict one peer. Since inbound eviction is triggered only in `CreateNodeFromAcceptedSocket()` (and not through some regularly scheduled job), this could lead to us having an extra inbound peer above our limit indefinitely. 
Now if the attacker repeats this scheme many times, they could make thousands of inbound connections to us / exhaust our resources / crash our node.

Comment (mzumsande) [reply src/net.cpp:907]: hmm, thinking more about this, I didn't see the whole picture, because everything happens in `ThreadSocketHandler` - after ` CConnman::SocketHandler() -> CConnman::SocketHandlerListening()`, `CConnman::DisconnectNodes()` is called, so it's not possible to accept new connections in between - however,  `CConnman::SocketHandlerListening()` calls `AcceptConnection()` in a loop for multiple `ListenSocket`s, so if we have multiple listening sockets and an attacker would connect to more than one at the same time, the scenario above may still be possible?

**Comment by dergoegge** [comment]
Rebased, have not addressed Martin's comments yet

**Comment by fanquake** [comment]
@instagibbs
